File: app/src/main/java/com/wafflestudio/snutt_staging/manager/TableManager.java
package com.wafflestudio.snutt_staging.manager;

import android.util.Log;

import com.wafflestudio.snutt_staging.SNUTTApplication;
import com.wafflestudio.snutt_staging.model.Coursebook;
import com.wafflestudio.snutt_staging.model.Table;

import java.util.ArrayList;
import java.util.Collections;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import retrofit.Callback;
import retrofit.RetrofitError;
import retrofit.client.Response;

/**
 * Created by makesource on 2016. 1. 16..
 */
public class TableManager {

    private static final String TAG = "TABLE_MANAGER" ;

    private Table current;
    private List<Table> tables;
    private Map<String, Table> tableMap;
    private SNUTTApplication app;

    private static TableManager singleton;

    /**
     * TableManager 싱글톤
     */

    private TableManager(SNUTTApplication app) {
        this.app = app;
        this.tables = new ArrayList<>();
        this.tableMap = new HashMap<>();
    }

    public static TableManager getInstance(SNUTTApplication app) {
        if(singleton == null) {
            singleton = new TableManager(app);
        }
        return singleton;
    }

    public static TableManager getInstance() {
        if (singleton == null) Log.e(TAG, "This method should not be called at this time!!");
        return singleton;
    }

    public void getTableList(final Callback callback) {
        String token = PrefManager.getInstance().getPrefKeyXAccessToken();
        app.getRestService().getTableList(token, new Callback<List<Table>>() {
            @Override
            public void success(List<Table> table_list, Response response) {
                Log.d(TAG, "get table list request success!");
                tables.clear();
                tableMap.clear();
                for (Table table : table_list) addTable(table);
                if (callback != null) callback.success(tables, response);
            }
            @Override
            public void failure(RetrofitError error) {
                Log.e(TAG, "get table list request failed..!");
                if (callback != null) callback.failure(error);
            }
        });
    }

    public void postTable(int year, int semester, String title, final Callback<List<Table>> callback) {
        String token = PrefManager.getInstance().getPrefKeyXAccessToken();
        Map query = new HashMap();
        query.put("year", year);
        query.put("semester", semester);
        query.put("title", title);
        app.getRestService().postTable(token, query, new Callback<List<Table>>() {
            @Override
            public void success(List<Table> table_list, Response response) {
                Log.d(TAG, "post new table request success!!");
                tables.clear();
                tableMap.clear();
                for (Table table : table_list) addTable(table);
                if (callback != null) callback.success(tables, response);
            }
            @Override
            public void failure(RetrofitError error) {
                Log.e(TAG, "post new table request failed..!");
                if (callback != null) callback.failure(error);
            }
        });
    }

    /*public void setTableList(List<Table> tables) {
        this.tables = tables;
    }*/

    public void getTableById(String id, final Callback callback) {
        String token = PrefManager.getInstance().getPrefKeyXAccessToken();
        app.getRestService().getTableById(token, id, new Callback<Table>() {
            @Override
            public void success(Table table, Response response) {
                Log.d(TAG, "get table by id success");
                LectureManager.getInstance().setLectures(table.getLecture_list());
                PrefManager.getInstance().updateNewTable(table);
                TagManager.getInstance().updateNewTag(table.getYear(), table.getSemester());
                if (callback != null) callback.success(table, response);
            }
            @Override
            public void failure(RetrofitError error) {
                Log.d(TAG, "get table by id is failed!");
                if (callback != null) callback.failure(error);
            }
        });
    }

    public void getDefaultTable(final Callback<Table> callback) {
        String token = PrefManager.getInstance().getPrefKeyXAccessToken();
        app.getRestService().getRecentTable(token, new Callback<Table>() {
            @Override
            public void success(Table table, Response response) {
                Log.d(TAG, "get recent table request success");
                LectureManager.getInstance().setLectures(table.getLecture_list());
                PrefManager.getInstance().updateNewTable(table);
                TagManager.getInstance().updateNewTag(table.getYear(), table.getSemester());
                if (callback != null) callback.success(table, response);
            }
            @Override
            public void failure(RetrofitError error) {
                Log.d(TAG, "get recent table request failed!");
                if (callback != null) callback.failure(error);
            }
        });
    }

    public void getCoursebook(final Callback<List<Coursebook>> callback) {
        //String token = PrefManager.getInstance().getPrefKeyXAccessToken();
        app.getRestService().getCoursebook(new Callback<List<Coursebook>>() {
            @Override
            public void success(List<Coursebook> coursebooks, Response response) {
                Log.d(TAG, "get coursebook request success.");
                if (callback != null) callback.success(coursebooks, response);
            }

            @Override
            public void failure(RetrofitError error) {
                Log.d(TAG, "get coursebook request failed.");
                if (callback != null) callback.failure(error);
            }
        });
    }

    public void addTable(Table table) {
        tables.add(table);
        tableMap.put(table.getId(), table);
        Collections.sort(tables);
    }

    public void updateTables(Table table) {
        // TODO : (SeongWon) server에 update 요청 날리기
    }
}

File: app/src/main/java/com/wafflestudio/snutt2/manager/TableManager.kt
package com.wafflestudio.snutt2.manager

import android.util.Log
import com.wafflestudio.snutt2.data.SNUTTStorage
import com.wafflestudio.snutt2.lib.network.SNUTTRestApi
import com.wafflestudio.snutt2.lib.network.dto.PostTableParams
import com.wafflestudio.snutt2.lib.network.dto.PutTableParams
import com.wafflestudio.snutt2.lib.network.dto.core.CourseBookDto
import com.wafflestudio.snutt2.lib.network.dto.core.TableDto
import io.reactivex.rxjava3.core.Single
import io.reactivex.rxjava3.schedulers.Schedulers
import javax.inject.Inject
import javax.inject.Singleton

/**
 * Created by makesource on 2016. 1. 16..
 */
@Singleton
class TableManager @Inject constructor(
    private val snuttRestApi: SNUTTRestApi,
    private val lectureManager: LectureManager,
    private val tagManager: TagManager,
    private val prefStorage: PrefStorage,
    private val storage: SNUTTStorage
) {
    private var tableMap: MutableMap<String, TableDto> = HashMap()

    fun reset() {
        tableMap = HashMap()
    }

    fun hasTimetables(): Boolean {
        return false
    }

    fun getTableList(): Single<List<TableDto>> {
        return snuttRestApi.getTableList()
            .subscribeOn(Schedulers.io())
            .doOnSuccess { result ->
                tableMap.clear()
                storage.tables.setValue(result)
                result.forEach { addTable(it) }
            }
            .doOnError {
                Log.e(TAG, "get table list request failed..!")
            }
    }

    fun postTable(year: Long, semester: Long, title: String?): Single<List<TableDto>> {
        return snuttRestApi.postTable(
            PostTableParams(year, semester, title)
        )
            .subscribeOn(Schedulers.io())
            .doOnSuccess {
                tableMap.clear()
                it.forEach { addTable(it) }
            }
            .doOnError {
                Log.e(TAG, "post new table request failed..!")
            }
    }

    fun getTableById(id: String?): Single<TableDto> {
        return snuttRestApi.getTableById(
            id!!
        )
            .subscribeOn(Schedulers.io())
            .doOnSuccess {
                Log.d(TAG, "get table by id success")
                lectureManager.setLectures(it.lectureList)
                lectureManager.clearSearchedLectures()
                prefStorage.updateNewTable(it)
                tagManager.updateNewTag(it.year.toInt(), it.semester.toInt())
            }
            .doOnError {
                Log.d(TAG, "get table by id is failed!")
            }
    }

    fun getDefaultTable(): Single<TableDto> {
        return snuttRestApi.getRecentTable()
            .subscribeOn(Schedulers.io())
            .doOnSuccess {
                Log.d(TAG, "get recent table request success")
                lectureManager.setLectures(it.lectureList)
                lectureManager.clearSearchedLectures()
                prefStorage.updateNewTable(it)
                tagManager.updateNewTag(it.year.toInt(), it.semester.toInt())
            }
    }

    fun getCoursebook(): Single<List<CourseBookDto>> {
        // String token = PrefManager.getInstance().getPrefKeyXAccessToken();
        return snuttRestApi.getCoursebook()
            .subscribeOn(Schedulers.io())
            .doOnSuccess {
                storage.courseBooks.setValue(it)
            }
    }

    fun deleteTable(id: String?): Single<List<TableDto>> {
        return snuttRestApi.deleteTable(
            id!!
        )
            .doOnSuccess {
                Log.d(TAG, "delete table request success.")
                tableMap.clear()
                for (table in it) addTable(table)
            }
            .doOnError {
                Log.d(TAG, "delete table request failed.")
            }
    }

    fun putTable(id: String?, title: String?): Single<List<TableDto>> {
        return snuttRestApi.putTable(
            id!!,
            PutTableParams(title!!)
        )
            .subscribeOn(Schedulers.io())
            .doOnSuccess {
                Log.d(TAG, "delete table request success.")
                tableMap.clear()
                for (table in it) addTable(table)
            }
            .doOnError {
                Log.d(TAG, "delete table request failed.")
            }
    }

    fun addTable(table: TableDto) {
        tableMap[table.id] = table
    }

    fun getTableTitleById(id: String): String? {
        val table = tableMap[id]
        if (table == null) {
            Log.e(TAG, "invalid table id..")
            return null
        }
        return table.title
    }

    fun updateTables(table: TableDto?) {
        // TODO : (SeongWon) server에 update 요청 날리기
    }

    companion object {
        private const val TAG = "TABLE_MANAGER"
    }
}
